src/porepy/numerics/nonlinear/nonlinear_solvers.py

```diff
@@ -165,15 +169,26 @@ def newton_step() -> None:
         if not is_converged:
             model.after_nonlinear_failure(sol, errors, iteration_counter)
 
-        return error_norm, is_converged, iteration_counter
+        return errors, is_converged, iteration_counter
 
-    def iteration(self, model) -> np.ndarray:
+    def iteration(self, model) -> tuple[np.ndarray, np.ndarray]:
         """A single nonlinear iteration.
 
         Right now, this is an almost trivial function. However, we keep it as a separate
         function to prepare for possible future introduction of more advanced schemes.
 
+        Parameters:
+            model: The model instance specifying the problem to be solved.
+
+        Returns:
+            The method returns the following tuple:
+
+            np.ndarray: Residual vector of nonlinear system.
+            np.ndarray: Solution to linearized system, i.e. the Newton update
+                increment.
+
         """
         model.assemble_linear_system()
+        _, res = model.linear_system  # Extract residual
         sol = model.solve_linear_system()
-        return sol
+        return res, sol
```

Comment:
It's incorrect to use this residual, collected before solving, for the error computation. We need to re-evaluate after solving and updating the solution.

Reply:
Yes, I realized this was not entirely the right way to do it. I am guessing we do not want to assemble the same system twice, though? So we need to assemble the system for extracting the residual, and then find a way to reuse this system for the next Newton iteration?

Reply:
This comment has been integrated in the recent changes. The residual is considered after update.